### PR TITLE
Update to b2646

### PR DESCRIPTION
**Pull request by mikebirdgeneau**
Update to b2646

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by mikebirdgeneau** [comment]
@conda-forge-admin, please rerender

**Comment by mikebirdgeneau** [comment]
Ok, so it looks like there's a little bit more going on here that I'm not familiar with:

1. Builds with CUDA versions other than None are failing. 
2. OSX build has `osx-64-pick-discrete.patch` which needs to be updated based on upstream changes.

I'm unfortunately not familiar enough with conda-forge builds (or the history associated with that patch) to carry this much further.

**Comment by frankier** [comment]
Thanks for this. It looks like you're almost done.

As far as I can see, there should not be a need for the Metal patch. If the patch is an improvement on picking the correct device, surely it belongs upstream in llama.cpp.

The default should be the discrete GPU anyway. As far as I can tell the patch just tries to pick this when available.

From https://developer.apple.com/documentation/metal/gpu_devices_and_work_submission/getting_the_default_gpu "On macOS devices that have multiple GPUs, such as a MacBook Pro, the system default is the discrete GPU."

Let's see about the CUDA overlinking...

**Comment by frankier** [comment]
It looks like the issue is a false error triggered by a new linting step which checks we haven't linked too much stuff into our libraries.

The solution is here:

https://conda-forge.org/docs/maintainer/knowledge_base/#nvcudadll-cannot-be-found-on-windows

Since it's happening on Linux too, I guess we could try:

```
build:
  ...
  missing_dso_whitelist:
    - "*/nvcuda.dll"   # [win]
    - "*/libcuda.so.1" # [linux]
```

@mikebirdgeneau Does this help? Do you want to fix up this PR now or should I open another one?

**Comment by mikebirdgeneau** [comment]
@conda-forge-admin, please rerender

**Comment by frankier** [comment]
I'm not entirely sure if this is correct, but it looks like someone solved a similar issue by only running tests on cpu. This makes sense since the build machine does not have Nvidia drivers, so cannot run anything with CUDA, only compile it.

Related issue: https://github.com/pytorch/audio/issues/488
Fix: https://github.com/pytorch/audio/pull/489/commits/4ee0db55b3a9ad1d2cddb7ff4153091f24a26acd

**Comment by mikebirdgeneau** [comment]
@frankier  I'm not sure either.  When I was looking at this last week, I compared to the build from December (last release) that worked just fine, without setting tests to run on CPU only.  

I concluded that the reason this is failing was because of updates in the conda-forge build / configuration, and that someone with a bit more experience with conda-forge, and CUDA might be better suited to look at this particular update, even though I'm happy to give it a go...

**Comment by mikebirdgeneau** [comment]
This might be a good configuration to look at `tensorflow-feedstock` for comparison: https://github.com/conda-forge/tensorflow-feedstock/blob/main/recipe/meta.yaml

**Comment by frankier** [comment]
I made a few changes here https://github.com/conda-forge/llama.cpp-feedstock/pull/24/files/9202c92f841392ab3e2295c8c60563b5d9f896d2..ff846d1a5683bf3cbec0f50c21be90f6b444b09b

Not all are necessary. What works is just disabling tests when building with CUDA.

I don't think comparing with Tensorflow is necessarily that informative. It looks like these executable are using dynamic linking, but I believe that something like a Python library will use dynamic loading, so that if the CUDA driver is not available, it will not immediately cause an error. In fact if there is a load time error, it may be that the library works around it (falls back to CPU) or masks it until later.

The reason that compilation is able to complete successfully is that the CUDA driver `cuda.so.1` is supplied as a stub file during compilation. It may be possible to "trick" the executable into running by supplying these stubs at test time too, given no CUDA functionality is used by printing the help message. The easiest thing to do is to just skip the tests, which really don't test very much.

I have build a linux x86_64 CUDA 12 package locally, and am able to install it and run it on a server with a GPU including with the `--ngl` flag.

As for why these changes are needed versus the older version, I think the dpo_whitelist is a result of updates to conda-forge's CI infastructure (i.e. what happens when you type "please rerender"). The tests needing `cuda.so.1` could be for different reasons. Perhaps llama.cpp is perhaps loading CUDA stuff earlier now? Perhaps to give more detailed diagnostics about GPUs? Or perhaps the stubs were previously being loaded during test time?
